### DISCUSSION TITLE
Source File Reorganization

### DISCUSSION THREAD

**Discussion by Veratil** [LMMS/lmms#5944] [Project Management]
Discussion for the subtopic of source file reorganization in #5592.

`include/` is a big target, we should aim for both `include/` and `src/` to have the same folder structure. Additionally, we can comment on namespacing classes.

I tried to make the *huge* list of files a little more managable (and less scroll heavy) by grouping it into collapsable groups based on folder structure. Let me know if we'd rather just have the `tree` output as-is instead.

<details>
<summary><code>/</code></summary>

<details>
<summary><code>include/</code></summary>

```
include/
├── AboutDialog.h
├── ActionGroup.h
├── aeffectx.h
├── AudioAlsa.h
├── AudioAlsaSetupWidget.h
├── AudioDevice.h
├── AudioDeviceSetupWidget.h
├── AudioDummy.h
├── AudioFileDevice.h
├── AudioFileFlac.h
├── AudioFileMP3.h
├── AudioFileOgg.h
├── AudioFileWave.h
├── AudioJack.h
├── AudioOss.h
├── AudioPortAudio.h
├── AudioPort.h
├── AudioPulseAudio.h
├── AudioSampleRecorder.h
├── AudioSdl.h
├── AudioSndio.h
├── AudioSoundIo.h
├── AutomatableButton.h
├── AutomatableModel.h
├── AutomatableModelView.h
├── AutomatableSlider.h
├── AutomationEditor.h
├── AutomationNode.h
├── AutomationPattern.h
├── AutomationPatternView.h
├── AutomationTrack.h
├── BandLimitedWave.h
├── base64.h
├── BasicFilters.h
├── BBEditor.h
├── BBTrackContainer.h
├── BBTrack.h
├── BufferManager.h
├── CaptionMenu.h
├── Clipboard.h
├── ColorChooser.h
├── ComboBox.h
├── ComboBoxModel.h
├── ConfigManager.h
├── ControlLayout.h
├── ControllerConnectionDialog.h
├── ControllerConnection.h
├── ControllerDialog.h
├── Controller.h
├── ControllerRackView.h
├── ControllerView.h
├── Controls.h
├── CPULoadWidget.h
├── CustomTextKnob.h
├── DataFile.h
├── debug.h
├── Delay.h
├── denormals.h
├── DeprecationHelper.h
├── DetuningHelper.h
├── DrumSynth.h
├── DspEffectLibrary.h
├── DummyEffect.h
├── DummyInstrument.h
├── DummyPlugin.h
├── Editor.h
├── EffectChain.h
├── EffectControlDialog.h
├── EffectControls.h
├── Effect.h
├── EffectRackView.h
├── EffectSelectDialog.h
├── EffectView.h
├── embed.h
├── endian_handling.h
├── Engine.h
├── EnvelopeAndLfoParameters.h
├── EnvelopeAndLfoView.h
├── ExportFilter.h
├── ExportProjectDialog.h
├── FadeButton.h
├── Fader.h
├── fenv.h
├── fft_helpers.h
├── FifoBuffer.h
├── FileBrowser.h
├── FileDialog.h
├── FxLine.h
├── FxLineLcdSpinBox.h
├── FxMixer.h
├── FxMixerView.h
├── Graph.h
├── GroupBox.h
├── GuiApplication.h
├── gui_templates.h
├── ImportFilter.h
├── InlineAutomation.h
├── InstrumentFunctions.h
├── InstrumentFunctionViews.h
├── Instrument.h
├── InstrumentMidiIOView.h
├── InstrumentPlayHandle.h
├── InstrumentSoundShaping.h
├── InstrumentSoundShapingView.h
├── InstrumentTrack.h
├── InstrumentView.h
├── interpolation.h
├── IoHelper.h
├── JournallingObject.h
├── Knob.h
├── Ladspa2LMMS.h
├── LadspaBase.h
├── LadspaControl.h
├── LadspaControlView.h
├── ladspa.h
├── LadspaManager.h
├── LcdSpinBox.h
├── LcdWidget.h
├── LedCheckbox.h
├── LeftRightNav.h
├── LfoController.h
├── LinkedModelGroups.h
├── LinkedModelGroupViews.h
├── lmms_basics.h
├── lmms_constants.h
├── lmms_math.h
├── LmmsPalette.h
├── LmmsStyle.h
├── LocaleHelper.h
├── LocklessAllocator.h
├── LocklessList.h
├── LocklessRingBuffer.h
├── Lv2Basics.h
├── Lv2ControlBase.h
├── Lv2Evbuf.h
├── Lv2Features.h
├── Lv2Manager.h
├── Lv2Options.h
├── Lv2Ports.h
├── Lv2Proc.h
├── Lv2SubPluginFeatures.h
├── Lv2UridCache.h
├── Lv2UridMap.h
├── Lv2ViewBase.h
├── MainApplication.h
├── MainWindow.h
├── MemoryHelper.h
├── MemoryManager.h
├── MeterDialog.h
├── MeterModel.h
├── MicroTimer.h
├── MidiAlsaRaw.h
├── MidiAlsaSeq.h
├── MidiApple.h
├── MidiCCRackView.h
├── MidiClient.h
├── MidiController.h
├── MidiDummy.h
├── MidiEvent.h
├── MidiEventProcessor.h
├── MidiEventToByteSeq.h
├── Midi.h
├── MidiJack.h
├── MidiOss.h
├── MidiPort.h
├── MidiPortMenu.h
├── MidiSetupWidget.h
├── MidiSndio.h
├── MidiWinMM.h
├── Mixer.h
├── MixerProfiler.h
├── MixerWorkerThread.h
├── MixHelpers.h
├── Model.h
├── ModelView.h
├── ModelVisitor.h
├── Note.h
├── NotePlayHandle.h
├── NStateButton.h
├── Oscillator.h
├── Oscilloscope.h
├── OutputSettings.h
├── panning_constants.h
├── panning.h
├── PathUtil.h
├── Pattern.h
├── PeakController.h
├── PerfLog.h
├── Piano.h
├── PianoRoll.h
├── PianoView.h
├── Pitch.h
├── PixmapButton.h
├── PlayHandle.h
├── PluginBrowser.h
├── PluginFactory.h
├── Plugin.h
├── PluginIssue.h
├── PluginView.h
├── PositionLine.h
├── PresetPreviewPlayHandle.h
├── ProjectJournal.h
├── ProjectNotes.h
├── ProjectRenderer.h
├── ProjectVersion.h
├── RecentProjectsMenu.h
├── RemotePlugin.h
├── RenameDialog.h
├── RenderManager.h
├── RingBuffer.h
├── RmsHelper.h
├── RowTableView.h
├── Rubberband.h
├── SampleBuffer.h
├── SamplePlayHandle.h
├── SampleRecordHandle.h
├── SampleTrack.h
├── SendButtonIndicator.h
├── SerializingObject.h
├── SetupDialog.h
├── shared_object.h
├── SideBar.h
├── SideBarWidget.h
├── SongEditor.h
├── Song.h
├── stdshims.h
├── StepRecorder.h
├── StepRecorderWidget.h
├── StringPairDrag.h
├── SubWindow.h
├── SweepOscillator.h
├── TabBar.h
├── TabButton.h
├── TabWidget.h
├── TemplatesMenu.h
├── TempoSyncKnob.h
├── TempoSyncKnobModel.h
├── TextFloat.h
├── ThreadableJob.h
├── TimeDisplayWidget.h
├── TimeLineWidget.h
├── TimePos.h
├── ToolButton.h
├── ToolPlugin.h
├── ToolPluginView.h
├── ToolTip.h
├── TrackContainer.h
├── TrackContainerView.h
├── TrackContentObject.h
├── TrackContentObjectView.h
├── TrackContentWidget.h
├── Track.h
├── TrackLabelButton.h
├── TrackOperationsWidget.h
├── TrackRenameLineEdit.h
├── TrackView.h
├── ValueBuffer.h
├── VersionedSaveDialog.h
├── versioninfo.h
├── volume.h
├── VstSyncController.h
└── VstSyncData.h
```

</details>

<details>
<summary><code>src/</code> excluding <code>3rdparty</code></summary>

```
src/
├── CMakeLists.txt
├── lmmsconfig.h.in
├── lmmsversion.h.in
```

<details>
<summary><code>src/core/</code></summary>
<details>
<summary><code>src/core/audio/</code></summary>

```
├── core
│   ├── audio
│   │   ├── AudioAlsa.cpp
│   │   ├── AudioDevice.cpp
│   │   ├── AudioFileDevice.cpp
│   │   ├── AudioFileFlac.cpp
│   │   ├── AudioFileMP3.cpp
│   │   ├── AudioFileOgg.cpp
│   │   ├── AudioFileWave.cpp
│   │   ├── AudioJack.cpp
│   │   ├── AudioOss.cpp
│   │   ├── AudioPortAudio.cpp
│   │   ├── AudioPort.cpp
│   │   ├── AudioPulseAudio.cpp
│   │   ├── AudioSampleRecorder.cpp
│   │   ├── AudioSdl.cpp
│   │   ├── AudioSndio.cpp
│   │   └── AudioSoundIo.cpp
```
</details>
<details>
<summary><code>src/core/lv2/</code></summary>

```
├── core
│   ├── lv2
│   │   ├── Lv2Basics.cpp
│   │   ├── Lv2ControlBase.cpp
│   │   ├── Lv2Evbuf.cpp
│   │   ├── Lv2Features.cpp
│   │   ├── Lv2Manager.cpp
│   │   ├── Lv2Options.cpp
│   │   ├── Lv2Ports.cpp
│   │   ├── Lv2Proc.cpp
│   │   ├── Lv2SubPluginFeatures.cpp
│   │   ├── Lv2UridCache.cpp
│   │   └── Lv2UridMap.cpp
```
</details>
<details>
<summary><code>src/core/midi/</code></summary>

```
├── core
│   ├── midi
│   │   ├── MidiAlsaRaw.cpp
│   │   ├── MidiAlsaSeq.cpp
│   │   ├── MidiApple.cpp
│   │   ├── MidiClient.cpp
│   │   ├── MidiController.cpp
│   │   ├── MidiEventToByteSeq.cpp
│   │   ├── MidiJack.cpp
│   │   ├── MidiOss.cpp
│   │   ├── MidiPort.cpp
│   │   ├── MidiSndio.cpp
│   │   └── MidiWinMM.cpp
```
</details>

```
├── core
│   ├── AutomatableModel.cpp
│   ├── AutomationNode.cpp
│   ├── AutomationPattern.cpp
│   ├── BandLimitedWave.cpp
│   ├── base64.cpp
│   ├── BBTrackContainer.cpp
│   ├── BufferManager.cpp
│   ├── Clipboard.cpp
│   ├── CMakeLists.txt
│   ├── ComboBoxModel.cpp
│   ├── ConfigManager.cpp
│   ├── ControllerConnection.cpp
│   ├── Controller.cpp
│   ├── DataFile.cpp
│   ├── DrumSynth.cpp
│   ├── EffectChain.cpp
│   ├── Effect.cpp
│   ├── Engine.cpp
│   ├── EnvelopeAndLfoParameters.cpp
│   ├── fft_helpers.cpp
│   ├── FxMixer.cpp
│   ├── ImportFilter.cpp
│   ├── InlineAutomation.cpp
│   ├── Instrument.cpp
│   ├── InstrumentFunctions.cpp
│   ├── InstrumentPlayHandle.cpp
│   ├── InstrumentSoundShaping.cpp
│   ├── JournallingObject.cpp
│   ├── Ladspa2LMMS.cpp
│   ├── LadspaControl.cpp
│   ├── LadspaManager.cpp
│   ├── LfoController.cpp
│   ├── LinkedModelGroups.cpp
│   ├── LocklessAllocator.cpp
│   ├── main.cpp
│   ├── MemoryHelper.cpp
│   ├── MemoryManager.cpp
│   ├── MeterModel.cpp
│   ├── MicroTimer.cpp
│   ├── Mixer.cpp
│   ├── MixerProfiler.cpp
│   ├── MixerWorkerThread.cpp
│   ├── MixHelpers.cpp
│   ├── Model.cpp
│   ├── ModelVisitor.cpp
│   ├── Note.cpp
│   ├── NotePlayHandle.cpp
│   ├── Oscillator.cpp
│   ├── PathUtil.cpp
│   ├── PeakController.cpp
│   ├── PerfLog.cpp
│   ├── Piano.cpp
│   ├── PlayHandle.cpp
│   ├── Plugin.cpp
│   ├── PluginFactory.cpp
│   ├── PluginIssue.cpp
│   ├── PresetPreviewPlayHandle.cpp
│   ├── ProjectJournal.cpp
│   ├── ProjectRenderer.cpp
│   ├── ProjectVersion.cpp
│   ├── RemotePlugin.cpp
│   ├── RenderManager.cpp
│   ├── RingBuffer.cpp
│   ├── SampleBuffer.cpp
│   ├── SamplePlayHandle.cpp
│   ├── SampleRecordHandle.cpp
│   ├── SerializingObject.cpp
│   ├── Song.cpp
│   ├── StepRecorder.cpp
│   ├── TempoSyncKnobModel.cpp
│   ├── TimePos.cpp
│   ├── ToolPlugin.cpp
│   ├── TrackContainer.cpp
│   ├── TrackContentObject.cpp
│   ├── Track.cpp
│   ├── ValueBuffer.cpp
│   └── VstSyncController.cpp
```
</details>
<details>
<summary><code>src/gui/</code></summary>

```
├── gui
│   ├── AboutDialog.cpp
│   ├── ActionGroup.cpp
│   ├── AudioAlsaSetupWidget.cpp
│   ├── AudioDeviceSetupWidget.cpp
│   ├── AutomatableModelView.cpp
│   ├── AutomationPatternView.cpp
│   ├── CMakeLists.txt
│   ├── ControllerConnectionDialog.cpp
│   ├── ControllerDialog.cpp
│   ├── EffectControlDialog.cpp
│   ├── EffectSelectDialog.cpp
│   ├── embed.cpp
│   ├── ExportProjectDialog.cpp
│   ├── FileBrowser.cpp
│   ├── FxMixerView.cpp
│   ├── GuiApplication.cpp
│   ├── InstrumentView.cpp
│   ├── LfoControllerDialog.cpp
│   ├── LmmsPalette.cpp
│   ├── LmmsStyle.cpp
│   ├── Lv2ViewBase.cpp
│   ├── MainApplication.cpp
│   ├── MainWindow.cpp
│   ├── MidiCCRackView.cpp
│   ├── MidiSetupWidget.cpp
│   ├── ModelView.cpp
│   ├── PeakControllerDialog.cpp
│   ├── PianoView.cpp
│   ├── PluginBrowser.cpp
│   ├── RowTableView.cpp
│   ├── SetupDialog.cpp
│   ├── StringPairDrag.cpp
│   ├── SubWindow.cpp
│   ├── TimeLineWidget.cpp
│   ├── ToolPluginView.cpp
│   ├── TrackContainerView.cpp
│   ├── TrackContentObjectView.cpp
│   ├── TrackView.cpp
```
<details>
<summary><code>src/gui/dialogs/</code></summary>

```
├── gui
│   ├── dialogs
│   │   ├── about_dialog.ui
│   │   ├── ColorChooser.cpp
│   │   ├── export_project.ui
│   │   ├── FileDialog.cpp
│   │   └── VersionedSaveDialog.cpp
```
</details>
<details>
<summary><code>src/gui/editors/</code></summary>

```
├── gui
│   ├── editors
│   │   ├── AutomationEditor.cpp
│   │   ├── BBEditor.cpp
│   │   ├── Editor.cpp
│   │   ├── PianoRoll.cpp
│   │   └── SongEditor.cpp
```
</details>
<details>
<summary><code>src/gui/Forms/</code></summary>

```
├── gui
│   ├── Forms
│   │   └── EffectSelectDialog.ui
```
</details>
<details>
<summary><code>src/gui/menus/</code></summary>

```
├── gui
│   ├── menus
│   │   ├── RecentProjectsMenu.cpp
│   │   └── TemplatesMenu.cpp
```
</details>
<details>
<summary><code>src/gui/widgets/</code></summary>

```
├── gui
│   └── widgets
│       ├── AutomatableButton.cpp
│       ├── AutomatableSlider.cpp
│       ├── CaptionMenu.cpp
│       ├── ComboBox.cpp
│       ├── ControlLayout.cpp
│       ├── ControllerRackView.cpp
│       ├── ControllerView.cpp
│       ├── Controls.cpp
│       ├── CPULoadWidget.cpp
│       ├── CustomTextKnob.cpp
│       ├── EffectRackView.cpp
│       ├── EffectView.cpp
│       ├── EnvelopeAndLfoView.cpp
│       ├── FadeButton.cpp
│       ├── Fader.cpp
│       ├── FxLine.cpp
│       ├── FxLineLcdSpinBox.cpp
│       ├── Graph.cpp
│       ├── GroupBox.cpp
│       ├── InstrumentFunctionViews.cpp
│       ├── InstrumentMidiIOView.cpp
│       ├── InstrumentSoundShapingView.cpp
│       ├── Knob.cpp
│       ├── LadspaControlView.cpp
│       ├── LcdSpinBox.cpp
│       ├── LcdWidget.cpp
│       ├── LedCheckbox.cpp
│       ├── LeftRightNav.cpp
│       ├── LinkedModelGroupViews.cpp
│       ├── MeterDialog.cpp
│       ├── MidiPortMenu.cpp
│       ├── NStateButton.cpp
│       ├── Oscilloscope.cpp
│       ├── PixmapButton.cpp
│       ├── PositionLine.cpp
│       ├── ProjectNotes.cpp
│       ├── RenameDialog.cpp
│       ├── Rubberband.cpp
│       ├── SendButtonIndicator.cpp
│       ├── SideBar.cpp
│       ├── SideBarWidget.cpp
│       ├── StepRecorderWidget.cpp
│       ├── TabBar.cpp
│       ├── TabWidget.cpp
│       ├── TempoSyncKnob.cpp
│       ├── TextFloat.cpp
│       ├── TimeDisplayWidget.cpp
│       ├── ToolButton.cpp
│       ├── ToolTip.cpp
│       ├── TrackContentWidget.cpp
│       ├── TrackLabelButton.cpp
│       ├── TrackOperationsWidget.cpp
│       └── TrackRenameLineEdit.cpp
```
</details>
</details>
<details>
<summary><code>src/tracks/</code></summary>

```
└── tracks
    ├── AutomationTrack.cpp
    ├── BBTrack.cpp
    ├── CMakeLists.txt
    ├── InstrumentTrack.cpp
    ├── Pattern.cpp
    └── SampleTrack.cpp
```
</details>
</details>
</details>

**Comment by Spekular** [comment]
One thing I find a bit confusing is the subdivisions in the GUI folder. Are the exactly defined and meaningful?
- There's a "dialogs" folder, but many "dialog" classes are directly in the GUI folder and one is under "forms".
- There's an "editors" folder, but inside LMMS these editors are at the same hierarchy level as the (not included) FX-Mixer, Controller Rack, and Project Notes.
- After the track classes are split, there will be many XyzTrackView classes, which should probably be in a single folder?

I also find the src/tracks folder rather weird. These should probably move in under core after the split. Plus, we have src/core/lv2, but no src/core/vst2 or src/core/ladspa. Are these two good candidates for new folders?

+1 for matching the src folder structure in the include directory.

**Reply by Veratil** [reply]
> One thing I find a bit confusing is the subdivisions in the GUI folder. Are the exactly defined and meaningful?

They are to an extent, but they need some management.

>     * There's a "dialogs" folder, but many "dialog" classes are directly in the GUI folder and one is under "forms".

We'll mark those for moving then.

>     * There's an "editors" folder, but inside LMMS these editors are at the same hierarchy level as the (not included) FX-Mixer, Controller Rack, and Project Notes.

We should move the mixer, rack, and notes under editors.

>     * After the track classes are split, there will be many XyzTrackView classes, which should probably be in a single folder?

That's a good idea, keeping them in a single spot would make things easier to find.

> I also find the src/tracks folder rather weird. These should probably move in under core after the split. Plus, we have src/core/lv2, but no src/core/vst2 or src/core/ladspa. Are these two good candidates for new folders?

Those sound good to me.

**Reply by DomClark** [reply]
> We have src/core/lv2, but no src/core/vst2 or src/core/ladspa. Are these two good candidates for new folders?

Most of the VST2 code is in `plugins/vst_base`. There's only a tiny amount of support code in the core, namely `VstSyncController`, which I don't think deserves its own folder.
The LV2 and LADSPA code is almost unused in the core - the only references are in `LmmsCore`, which holds a "manager" for each. These are responsible for loading and enumerating plugins, and are created at program startup and destroyed at program shutdown. I would instead like to move LV2 and LADSPA to their own plugins like VST2, e.g. `plugins/Lv2Base` and `plugins/LadspaBase`; the startup and shutdown handlers could be implemented via hooks as in #5215. (See my [WIP VST3 branch](https://github.com/DomClark/lmms/tree/vst3) for a [implementation of such hooks](https://github.com/DomClark/lmms/commit/adb81546520629a6258db1c3e66aab47abf25464), and a [plugin that uses them to create/destoy a "manager"](https://github.com/DomClark/lmms/blob/14a132ea128cd56ff1c5710fc91c3d39cde0f180/plugins/Vst3Base/Vst3Base.cpp).)

**Comment by JohannesLorenz** [comment]
Similar to what Spekular said about dialogs: In the *current* file tree, I often look for widget files and don't know if they are in `src/gui` or `src/gui/widgets`.

All in all, I'm not always for creating more folders. E.g. if you have a folder for "TrackView" classes, one for menus, one for editors etc., then when you look for a gui file, you always need to think "Is it a track view? No. Is it a menu? No. Is it an editor? No" etc. . I guess I'm faster just finding all in "src/gui" and scroll to the right starting letter in my IDE...

**Comment by M374LX** [comment]
With only one file, the folder ``src/gui/Forms`` is unnecessary. A better idea is to move ``EffectSelectDialog.ui`` to ``src/gui/dialogs``.

I also suggest moving ``main.cpp`` out of ``src/core`` and placing it directly in ``src``, as this would give more visibility to the source file in which the program starts. Additionally, ``main.cpp`` inevitably depends on stuff from ``src/gui``, even in an ideal world where ``src/core`` is independent from ``src/gui``.

**Reply by Veratil** [reply]
👍 for Forms folder.

I think `main.cpp` is okay where it's at for now. Eventually if we can split the core entirely away from the GUI, then we'd have something like `liblmms` which controls all the core stuff and then `lmms` is the GUI frontend.

**Reply by PhysSong** [reply]
For `.ui` files, we may move and rename them to leverage CMake's [`AUTOUIC`](https://cmake.org/cmake/help/latest/prop_tgt/AUTOUIC.html) feature which we don't use at this moment.

**Comment by DomClark** [comment]
Which classes should go in `src/gui/widgets`? I imagine the vast majority of GUI classes derive (not necessarily directly) from `QWidget`, so that can't be a sufficient criterion. Here's my proposal:

Qt has a number of classes that come in groups of three, e.g. `QAbstractListModel`, `QListView`, `QListWidget`. The `XView` class is designed to display data from the `AbstractXModel` class, but there's also a `XWidget` class that has its own built-in model, for situations where a model-view system isn't desirable. In LMMS, this system would make sense in some cases, but not others. For example, there's no use having a track view that doesn't represent an actual track in the project, so a `TrackWidget` that could be used alone isn't useful. However, there are plenty of cases where we use an `LedCheckbox` without it viewing some model from the project (config dialog, controller connection dialog, etc.). These generic widgets that can be reused across the UI are what I think should go in the `widgets` directory, whereas specialised single-purpose widgets should go elsewhere.

**Reply by Veratil** [reply]
Agreed. Any widget which is used across multiple features is a candidate for placement in the widgets folder. Knobs, Checkboxes, etc.

**Reply by PhysSong** [reply]
There are several kinds of GUI classes in LMMS, including but not limited to:
- Generic widgets which can be reused everywhere(`Knob`, `Fader`, etc.)
- Widgets which can be reused only in some contexts(`MixerLineLcdSpinBox`, etc.)
- Views which represents a specific model(or group) as-is(`MeterDialog`, etc.)
- Views which is only useful as part of a larger view(`InstrumentMidiIOView`, etc.)
- GUI elements which isn't associated to specific model(`Oscilloscope`, `CPULoadWidget`, etc.)
- Utility classes which represents neither a widget nor a view

I think we should decide which directory should be chosen for each cases.
Also, I suggest grouping related sub-views(e.g., `InstrumentMidiIOView`, `InstrumentMiscView`, `InstrumentSoundShapingView`, ...) in a folder.

**Reply by JohannesLorenz** [reply]
You could group them, but imagine you just want to open `Knob.cpp`. Then you type `src/gui/widgets/`, and then, instead of just typing `Kn<TAB>`, you need to think "hmm... which folder was it in" or type */Kn* (and pray this does not hit 0 or 2 files, because TAB completion does not work here). So isn't this a disadvantage?

**Reply by PhysSong** [reply]
I didn't mean grouping everything in addition to the current hierarchy. I meant grouping sub-views into a directory like `src/gui/instrument`.

**Reply by allejok96** [reply]
Like PhysSong said, it's hard to categorize widgets into "general" or "specific". There are many FileBrowser in the SideBar and TimeLineWidget is used in all editors. NStateButton and RowTableView feels very general but are only used once.

What if we divided it into *where* instead of *what*. So Oscilloscope would go into toolbar/ and so on, and things like LedCheckBox into widgets/

.... Or just put everything in the same directory. It's quicker to search by letter than trying to define what kind of thing you're searching for.

**Reply by allejok96** [reply]
When diving into to the LMMS codebase for the first time it can be very hard to figure out which cpp file contains the part of the GUI I'm interested in. So having directories for the parts with lots of custom widgets, like songeditor, mixer, and instrument(window) would help.

**Comment by PhysSong** [comment]
I think we should eventually separate GUI headers from core headers to a new directory(`include/gui`).
For now, it can't be done cleanly because some core classes depend on GUI functionalities.

**Comment by allejok96** [comment]
https://github.com/allejok96/lmms/wiki/Widget-overview

In the link above I've tried to categorize widgets based on where they are used, and what widgets they contain/are contained by.

|Category|Number of files|Group in sub-directory?|
|-|-|-|
Common widgets | 28 | yes
Controllers |  6 | maybe
Dialogs | 8 | yes
Editors | 10 | yes
Effects | 4 | 
Instrument | 8 | yes
LV2 | 4 | 
Non-widget classes | 8 |  
Main window | 8 | 
Midi | 2 | 
Mixer | 4 | 
Settings | 4 | maybe
Tracks (and clips) | 16 | yes
Other subwindows | 3 |

**Comment by tipico-games** [comment]
The proposal suggests aligning the folder structures of `include/` and `src/` for better organization and consistency. Key points:

1. **Goal**: Mirror `include/` and `src/` structures to improve maintainability.
2. **Namespacing**: Discuss proper class namespacing for clarity.
3. **Grouped View**: Files are grouped into collapsible folders for easier navigation and reduced scrolling.

Feedback is welcome on the folder structure and whether the grouped view works better than raw tree output.
